Ticket4045 revert motor velocity after move

ThomasLohnert opened the pull request:
### Description of work

Issuing a move involving multiple axes via the reflectometry server alters the velocity of each motor so that the move duration matches that of the slowest axis. The changes in this PR are to ensure that each motor reverts to its original velocity prior to the coordinated move once it is finished.

To Test:
- Using this example config (put in `<inst>/configurations/refl/config.py`) containing just a theta parameter and a detector with axes for height and angle. (Config assumes a GALIL IOC running and mapped to MTR01XX):
```
from ReflectometryServer import *

SPACING = 1000


def get_beamline():
    # Positions along beam
    z_sample = 1 * SPACING
    z_point_detector = 2 * SPACING

    # COMPONENTS
    detector = TiltingComponent("detector", PositionAndAngle(0.0, z_point_detector, 90))
    theta = ThetaComponent("theta", PositionAndAngle(0.0, z_sample, 90), [detector])

    comps = [theta, detector]

    # BEAMLINE PARAMETERS
    track_point_detector = TrackingPosition("track_point_detector", detector)

    angle_theta = AngleParameter("theta", theta)
    angle_point_detector = AngleParameter("angle_detector", detector)

    params = [angle_theta, track_point_detector, angle_point_detector]

    # DRIVERS
    drivers = [DisplacementDriver(detector, MotorPVWrapper("MOT:MTR0107")),
               AngleDriver(detector, MotorPVWrapper("MOT:MTR0108"))
               ]

    # MODES
    nr_mode = BeamlineMode("NR", [param.name for param in params], {})
    modes = [nr_mode]

    beam_start = PositionAndAngle(0.0, 0.0, 0.0)
    bl = Beamline(comps, params, drivers, modes, beam_start)

    return bl
```
- Write a value to `<INSTPREFIX>:REFL:PARAM:THETA:SP` (this will set off the beamline move)
- Motors 0107 and 0108 should move in a coordinated fashion (i.e. their velocities will change so that they finish at the same time).

### To test

https://github.com/ISISComputingGroup/IBEX/issues/4045

### Acceptance criteria

(as per original ticket):
- [x] If a coordinated move is completed, each motor's `.VELO` field will return to the value it had before the move started (not necessarily maximum velocity)
- [x] If a coordinated move is interrupted (e.g. `Stop All` button), each motor's `.VELO` field will return to the value it had before the move started (not necessarily maximum velocity)
- [x] If a coordinated move is started while another coordinated move is already in progress, the motor's `.VELO` field will return to the value it had before the initial move started (not the one it had before the second move started)
- [x] If a value is written to a motor's `.VELO` field directly by the user while a coordinated move is ongoing, the motor will retain this new value after the move finishes rather than returning to the original velocity.
- [X] On startup, the motors' cached velocity to recover it autosaved value  if exists or the current velocity if not. This is a simplification to avoid race conditions between PV monitors on startup. Happy to discuss if you disagree with this decision

---

#### Code Review

- [x] Is the code of an acceptable quality?
- [x] Has the author taken into account the multi-threaded nature of the code?
- [x] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [x] Has the [manual system tests spreadsheet](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Manual-system-tests) been updated?

### Functional Tests

- [x] Do changes function as described? Add comments below that describe the tests performed.

### Final steps

- [ ] Reviewer has updated the submodule in the main EPICS repo? See **Reviewing work for the subModules of EPICS** in the [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [x] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

DominicOram commented:
As discussed, can we autosave the previous velocity rather than rely on using the max velocity. This could be dangerous as things may unexpectedly speed up.

I'm also getting a race condition where the new velocity being set by the reflectometry server is the one being cached:
![image](https://user-images.githubusercontent.com/10086797/59869647-b0f38f80-938b-11e9-8d04-99f8856b81ca.png)

ThomasLohnert commented:
Done as part of rework:
- Eliminated race condition which resulted in velocity by a synchronized move being set as velocity to restore
- On IOC startup, cached velocity to restore is now initialised to current value if motor is stationary, or from autosave file if it is moving
- Fixed issues with restoring velocity on PVWrappers for driving Jaws
- Added some unit tests for PVWrappers
- Moved logic for restoring motor settings from IOC Driving layer down to the PVWrappers
- Refactored autosave mechanism to be more generic
- Refactored PVWrapper class structure

JawsPVWrappers differ from regular MotorPVWrappers in that they do not update their cached velocity to restore after a move when it is manually set on the low level motor pv. This helps avoid crash conditions as the reflectometry server always sets / restores the velocity of the individual jaw blades in one axis to the same velocity.
